fhir-path/src/main/java/com/ibm/fhir/path/function/ResolveFunction.java

```diff
@@ -94,16 +112,51 @@ public int getMaxArity() {
                             }
                         }
                     } else {
-                        Matcher matcher = REFERENCE_PATTERN.matcher(referenceReference);
-                        if (matcher.matches()) {
-                            resourceType = matcher.group(RESOURCE_TYPE_GROUP);
-                            if (referenceType != null && !resourceType.equals(referenceType)) {
-                                throw new IllegalArgumentException("Resource type found in reference URL does not match reference type");
+                        if (isBundleContext) {
+                            // we know the root of the tree is a Bundle and the current node is of type Reference,
+                            // so walk up the tree until we get to highest Bundle.entry.fullUrl in the tree
+                            // and save the value of its fullUrl
+                            String fullUrl = null;
+                            FHIRPathTree tree = evaluationContext.getTree();
+                            FHIRPathNode nodeUnderTest = tree.getParent(node);
+                            while (nodeUnderTest != root) {
+                                if (nodeUnderTest.isResourceNode()) {
+                                    FHIRPathNode sibling = tree.getSibling(nodeUnderTest, "fullUrl");
+                                    if (sibling != null && sibling.hasValue()) {
+                                        fullUrl = sibling.getValue().asStringValue().string();
+                                    }
+                                }
+                                nodeUnderTest = tree.getParent(nodeUnderTest);
                             }
-                            String baseUrl = matcher.group(BASE_URL_GROUP);
-                            if ((baseUrl == null ||  matchesServiceBaseUrl(baseUrl)) && evaluationContext.resolveRelativeReferences()) {
-                                // relative reference
-                                resource = resolveRelativeReference(evaluationContext, node, resourceType, matcher.group(LOGICAL_ID_GROUP), matcher.group(VERSION_ID_GROUP));
+
+                            String bundleReference = FHIRUtil.buildBundleReference(reference, fullUrl);
```

Comment:
Note for others reviewing... when fuller is null, it'll return the referenceUriString = ref.getReference().getValue();

when this condition is true (ref.getReference() != null && ref.getReference().getValue() != null)